Include locale specifier in WPCom API requests

kurzee opened the pull request:
Closes #3780 

To test:

- First launch the app as normal and generate some known error messages from the API, or perhaps breaking some request params in code.
- Change your simulator/device language in Settings > General > Language and Region > iPhone Language.
- Generate errors again, ensure the errors display in the selected language.

Note: Not all iOS languages are supported via WordPress.com, which will default itself to "en".

Needs review: @astralbodies

koke commented:
You might want to check [WordPressComLanguageDatabase](https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/WordPress/Classes/Utility/Languages.swift) 

You can see it in use in the (already removed) plans remotes https://github.com/wordpress-mobile/WordPress-iOS/blob/c0ea5d3f23272a861cd6b2132e52528285e4236f/WordPress/Classes/Networking/PlansRemote.swift#L12

koke commented on WordPress/WordPressApi/WordPressComRestApi.swift:
```diff
@@ -88,8 +94,10 @@ public class WordPressComRestApi: NSObject
                      success: SuccessResponseBlock,
                      failure: FailureReponseBlock) -> NSProgress?
     {
+        let URLString = appendLocaleParameterToURLPathIfNeeded(URLString)
```
Wouldn't it be much simpler to add the `locale` value to `parameters` instead of modifying the URL?

kurzee replied:
@koke has to be a part of the URL unfortunately 😐

kurzee replied:
For POST, that is*

kurzee commented:
>You might want to check WordPressComLanguageDatabase.deviceLanguage

This is great, thanks @koke!

kurzee commented:
Replaced the previous `NSLocale.preferredLanguages` with the new `WordPressComLanguageDatabase` slug.

astralbodies commented:
Can we also add this to `WordPressComOAuthClient` maybe? Another PR would be fine. 😄

kurzee commented:
I had originally left that one out since the error messages are parsed based on taking an auth action. But, just now I went ahead to see if the oauth endpoints even supported the `locale` parameter and they do not. I'd say we leave it out for now as I imagine that oauth errors or more-so for a developer than an a user.

astralbodies commented:
@kurzee That makes me weep uncontrollably.

astralbodies commented:
Just realized that stats won't gain the benefit of this because the client exists only in the main app and not in the shared repo. Fine for now; will need to make a similar update in stats and figure out how to share that language DB.

astralbodies commented:
Looks good! I've found a few areas where the user is never notified of a problem with sync (overridden site ID to `-1` didn't cause any visual indicator) but that's another issue. 😄 

:shipit:

astralbodies commented:
Just a quick thought - is it possible to write a unit test for the client? Maybe override the locale and verify the parameter is set properly?

kurzee commented:
>will need to make a similar update in stats and figure out how to share that language DB.

That probably makes sense anyways, and we could export this helper function for appending the correct parameter.

kurzee commented:
>Just a quick thought - is it possible to write a unit test for the client? Maybe override the locale and verify the parameter is set properly?

Thanks for the suggestion on this @astralbodies. I rethought what I was doing a bit with the locale parameter and changed things up. Can you take a peak at the new changes?

I moved the locale path appending logic as an extension method for `WordPressComRestApi`. It didn't make 100% sense to have this as an actual method within the class. I also added testing compatibility for the new parameter, **please run all tests!** And, let me know if you think any of this is 🔪

astralbodies commented:
@kurzee feel free to merge, my :shipit: still stands 😀

kurzee commented:
Thanks @astralbodies!
